From bf7faf2e93def91be1576f9005b02deafb4512cc Mon Sep 17 00:00:00 2001
From: Ryan Blue <blue@apache.org>
Date: Wed, 22 Apr 2015 14:53:38 -0700
Subject: [PATCH 052/115] CDK-898: Fix file not found bug in distributed cache.

Setup in the distributed cache is failing when using the local job
runner when HDFS is the default file system. This is very likely related
to HDFS-7031. The work-around is to set the local FS as the default when
using LocalJobRunner. This is safe because HDFS has already been loaded,
so lookups without the authority with still resolve to the full URI.
---
 .../main/java/org/kitesdk/tools/TransformTask.java |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/kite-tools-parent/kite-tools/src/main/java/org/kitesdk/tools/TransformTask.java b/kite-tools-parent/kite-tools/src/main/java/org/kitesdk/tools/TransformTask.java
index d11da1d..8b0a5cf 100644
--- a/kite-tools-parent/kite-tools/src/main/java/org/kitesdk/tools/TransformTask.java
+++ b/kite-tools-parent/kite-tools/src/main/java/org/kitesdk/tools/TransformTask.java
@@ -32,7 +32,9 @@ import org.apache.crunch.impl.mr.MRPipeline;
 import org.apache.crunch.types.PType;
 import org.apache.crunch.types.avro.AvroType;
 import org.apache.crunch.types.avro.Avros;
+import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.conf.Configured;
+import org.apache.hadoop.fs.FileSystem;
 import org.apache.hadoop.hive.conf.HiveConf;
 import org.kitesdk.compat.DynMethods;
 import org.kitesdk.data.Dataset;
@@ -93,7 +95,14 @@ public class TransformTask<S, T> extends Configured {
 
   public PipelineResult run() throws IOException {
     if (isLocal(from.getDataset()) || isLocal(to.getDataset())) {
-      getConf().set("mapreduce.framework.name", "local");
+      // copy to avoid making changes to the caller's configuration
+      Configuration conf = new Configuration(getConf());
+      conf.set("mapreduce.framework.name", "local");
+      // replace the default FS for Crunch to avoid distributed cache errors.
+      // this doesn't affect the source or target, they are fully-qualified.
+      conf.set("fs.defaultFS", "file:/");
+      conf.set("fs.default.name", "file:/");
+      setConf(conf);
     }
 
     PType<T> toPType = ptype(to);
-- 
1.7.0.4

